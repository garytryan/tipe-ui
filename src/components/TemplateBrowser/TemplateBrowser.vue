<template>
  <div :data-tipe-ui="$options.name">
    <tipe-template-browser-toolbar
      @filterChange="$emit('filter-change', $event.value)"
      @newTemplate="$emit('new-template', $event.value)"
    />
    <tipe-finder
      :files="templates"
      icon="details"
    />
  </div>
</template>

<script>
import vueTypes from 'vue-types'
import FileShape from '@/types/File'
import TipeFinder from '@/components/Finder'
import TipeTemplateBrowserToolbar from './TemplateBrowserToolbar.vue'

export default {
  name: 'TipeTemplateBrowser',
  components: {
    TipeFinder,
    TipeTemplateBrowserToolbar
  },
  props: {
    templates: vueTypes.arrayOf(vueTypes.shape(FileShape))
  }
}
</script>

<style>
[data-tipe-ui='TipeTemplateBrowser'] {
  height: 100%;
  width: 100%;
  display: grid;
  grid-auto-rows: 4rem 1fr;
  grid-template-columns: 1fr;
  grid-gap: 0.375rem;
}
</style>
