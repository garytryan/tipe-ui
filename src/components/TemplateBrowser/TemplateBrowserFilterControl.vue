<template>
  <div :data-tipe-ui="$options.name">
    <button
      v-for="({ label, value }, i) in filters"
      :key="i"
      :class="buttonClassObject(value)"
      @click="() => onClick(value)"
    >
      {{ label }}
    </button>
  </div>
</template>

<script>
const filters = [
  { label: 'Recently Viewed', value: 'recently-viewed' },
  { label: 'Starred', value: 'starred' },
  { label: 'Created by me', value: 'created-by-me' },
  { label: 'Shared with me', value: 'shared-with-me' }
]

export default {
  name: 'TipeTemplateBrowserFilterControl',
  data() {
    return {
      value: 'recently-viewed',
      filters
    }
  },
  methods: {
    buttonClassObject(value) {
      return {
        on: value === this.value
      }
    },
    onClick(value) {
      this.value = value
      this.$emit('change', value)
    }
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeTemplateBrowserFilterControl'] {
  height: auto;
  width: auto;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
}

button {
  height: auto;
  width: auto;
  border: none;
  outline: none;
  padding: 0 1rem 0 0;
  margin: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 0.8125rem;
  background-color: transparent;
  white-space: nowrap;
}

button.on {
  font-weight: 800;
}
</style>
