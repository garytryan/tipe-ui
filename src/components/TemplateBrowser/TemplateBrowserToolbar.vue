<template>
  <div :data-tipe-ui="$options.name">
    <div class="filter-control-container">
      <tipe-template-browser-filter-control
        @change="$emit('filter-change', $event.value)"
      />
    </div>
    <div class="new-template-button-container">
      <tipe-button
        icon-before="plus"
        @click="$emit('new-template')"
      >
        New Template
      </tipe-button>
    </div>
  </div>
</template>

<script>
import TipeButton from '@/components/Button'
import TipeTemplateBrowserFilterControl from './TemplateBrowserFilterControl'

export default {
  name: 'TipeTemplateBrowserToolbar',
  components: {
    TipeButton,
    TipeTemplateBrowserFilterControl
  }
}
</script>

<style lang="postcss" scoped>
[data-tipe-ui='TipeTemplateBrowserToolbar'] {
  height: 100%;
  width: 100%;
  display: grid;
  grid-template-columns: 1fr auto;
  grid-auto-flow: 100%;
}

.filter-control-container {
  display: flex;
  align-items: flex-end;
  justify-content: flex-start;
}

.new-template-button-container {
  display: flex;
  align-items: center;
  justify-content: center;
}
</style>
