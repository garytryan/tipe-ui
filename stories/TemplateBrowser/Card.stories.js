import { storiesOf } from '@storybook/vue'

import FolderCard from '@/components/FolderCard'
import DocCard from '@/components/DocCard'
import LayoutCard from '@/components/LayoutCard'

storiesOf('Cards', module)
  .add('folder card', () => ({
    components: { FolderCard },
    template: '<folder-card>Hero Sections</folder-card>'
  }))
  .add('layout card with text', () => ({
    components: { LayoutCard, FolderCard },
    template: '<layout-card><h1>Hello World</h1></layout-card>'
  }))
  .add('doc card default', () => ({
    components: { DocCard },
    template: '<doc-card>About</doc-card>'
  }))
  .add('doc card recent', () => ({
    components: { DocCard },
    template: '<doc-card recent>About</doc-card>'
  }))
