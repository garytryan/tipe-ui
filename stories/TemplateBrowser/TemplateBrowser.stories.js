import { storiesOf } from '@storybook/vue'
import TipeTemplateBrowser from '@/components/TemplateBrowser'
import fileMock from '@/mocks/File'
import createManyMocks from '@/mocks/createManyMocks'

const templates = createManyMocks(() => fileMock({ type: 'template' }), 30)

storiesOf('TemplateBrowser', module).add('default', () => ({
  components: { TipeTemplateBrowser },
  data() {
    return {
      templates
    }
  },
  template: `<tipe-template-browser :templates="templates"/>`
}))
