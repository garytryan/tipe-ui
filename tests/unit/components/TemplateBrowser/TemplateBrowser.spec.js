import { shallowMount } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import TemplateBrowser from '@/components/TemplateBrowser'
import Finder from '@/components/Finder'
import { mocks } from '@tipe/tipe-test-utils'

describe('TemplateBrowser', () => {
  describe('<TemplateBrowser>', () => {
    it('matches previous snapshot', () => {
      const renderer = createRenderer()
      const wrapper = shallowMount(TemplateBrowser)

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })

    it('has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(TemplateBrowser)

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeTemplateBrowser')
    })
  })

  describe(':props', () => {
    it(':templates - should pass templates to Finder files props', () => {
      const propsData = { templates: [mocks.file()] }
      const wrapper = shallowMount(TemplateBrowser, { propsData })

      expect(wrapper.find(Finder).props().files).toEqual(propsData.templates)
    })
  })
})
