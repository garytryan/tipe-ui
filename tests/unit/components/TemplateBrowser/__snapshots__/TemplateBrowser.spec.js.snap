// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateBrowser <TemplateBrowser> matches previous snapshot 1`] = `
<div data-tipe-ui="TipeTemplateBrowser">
  <!---->
  <!---->
</div>
`;
